File: 프로그래머스/unrated/133500. 등대/등대.java
import java.io.*;
import java.util.*;

class Solution {
    static int N;
    static int answer = 0;
    static boolean[] light;
    static ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
    public int solution(int n, int[][] lighthouse) {
        N = n;
        light = new boolean[N+1];
        for(int i = 0; i <= N; i++){
            graph.add(new ArrayList<>());
        }
        for(int i = 0; i < lighthouse.length; i++){
            int from = lighthouse[i][0];
            int to = lighthouse[i][1];
            graph.get(from).add(to);
            graph.get(to).add(from);
        }
        
        dfs(1,1);
        return answer;
    }
    private static void dfs(int cur, int parent){
        for(int i = 0; i < graph.get(cur).size(); i++){
            int next = graph.get(cur).get(i);
            if(next != parent){
                dfs(next, cur);
                if(!light[next] && !light[cur]){
                    light[cur] = true;
                    answer++;
                }
            }
        }
    }
}

File: 프로그래머스/3/133500. 등대/등대.cpp
#include <string>
#include <vector>

using namespace std;
bool light[100001];
vector<int> graph[100001];
int answer = 0;
void dfs(int node, int parent){
    for(int i = 0; i < graph[node].size(); i++){
        if(graph[node][i] != parent){
            dfs(graph[node][i], node);
            if(light[graph[node][i]] == false && light[node] == false){
                light[node] = true;
                answer++;
            }
        }
    }
}
int solution(int n, vector<vector<int>> lighthouse) {
    
    for(int i = 0; i < lighthouse.size(); i++){
        int v1 = lighthouse[i][0];
        int v2 = lighthouse[i][1];
        graph[v1].push_back(v2);
        graph[v2].push_back(v1);
    }
    
    dfs(1,1);
    return answer;
}